File: CPP/ByRound/1741/1741_D_Masha_and_a_Beautiful_Tree.cpp
// Problem: CF 1741 D - Masha and a Beautiful Tree
// https://codeforces.com/contest/1741/problem/D

/*
Algorithm/Technique: 
The approach is based on a recursive-like strategy applied level by level from bottom to top of the complete binary tree.
Each level corresponds to a step where we compare pairs of elements that are at a distance of k (which doubles each step).
If the elements are out of order (i.e., a[i] > a[i + k]), we swap them and increment the operation count.
We also check if there's an inversion in the arrangement using a helper condition:
abs(a[i] - a[i + k]) != k checks if elements are correctly spaced for a valid permutation in the current level.
If any inversion is detected, we return -1.

Time Complexity: O(m * log m)
Space Complexity: O(m)
*/
#include <stdlib.h>
#include <iostream>
#include <utility>

using namespace std;
int main() {
  cin.tie(0)->sync_with_stdio(0);
  int q;
  cin >> q;
  while (q--) {
    int m;
    cin >> m;
    int a[m];
    for (int &i : a)
      cin >> i;
    bool inv = 0;
    int cnt = 0;
    for (int k = 1; k < m; k <<= 1) {
      for (int i = 0; i < m; i += 2 * k) {
        // Check if the current pair of elements satisfies the condition to be in order
        if (abs(a[i] - a[i + k]) != k)
          inv = 1;
        // Swap if elements are out of order
        if (a[i] > a[i + k]) {
          swap(a[i], a[i + k]);
          cnt++;
        }
      }
    }
    cout << (inv ? -1 : cnt) << "\n";
  }
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByTier/D/1741_D_Masha_and_a_Beautiful_Tree.py
# Problem: CF 1741 D - Masha and a Beautiful Tree
# https://codeforces.com/contest/1741/problem/D

"""
Algorithm: Tree Sorting via Swap Operations
This solution determines the minimum number of swaps required to sort a binary tree 
represented as a permutation in a specific way. The key idea is to recursively check 
and fix the ordering of elements in pairs at each level of the tree.

Approach:
- The permutation represents the leaves of a complete binary tree.
- We process the tree level by level from bottom to top.
- For each level, we check if adjacent elements in pairs are in correct relative order.
- If not, we swap their positions and increment the operation count.
- If at any point, the required condition fails (the difference in values doesn't match the required gap), 
  then the tree cannot be made beautiful and we return -1.

Time Complexity: O(n log n) where n is the size of the permutation.
Space Complexity: O(1) excluding input storage.
"""

for _ in range(int(input())):
    n = int(input())
    p = [*map(int, input().split())]
    j, ans, f = 1, 0, 0
    while j < n:
        for i in range(0, n, j * 2):
            # Check if the absolute difference between elements in pair is correct
            if abs(p[i] - p[i + j]) != j:
                f = 1  # Mark impossibility
            # Swap if out of order
            if p[i] > p[i + j]:
                p[i], p[i + j], ans = p[i + j], p[i], ans + 1
        j <<= 1  # Move to next level (left shift by 1 = multiply by 2)
    print([ans, -1][f])


# https://github.com/VaHiX/CodeForces/